akka-actor-tests/src/test/scala/akka/actor/CoordinatedShutdownSpec.scala

```diff
@@ -150,8 +149,96 @@ class CoordinatedShutdownSpec
-      receiveN(4) should ===(List("A", "B", "B", "C"))
+      whenReady(co.run(UnknownReason)) { _ =>
+        receiveN(4) should ===(List("A", "B", "B", "C"))
+      }
+    }
+
+    "cancel tasks" in {
+      import system.dispatcher
+      val phases = Map("a" -> emptyPhase, "b" -> phase("a"))
+      val co = new CoordinatedShutdown(extSys, phases)
+
+      val task0Counter = new AtomicInteger()
+
+      val task0: () => Future[Done] = () =>
+        Future {
+          task0Counter.incrementAndGet()
+          Done
+        }
+
+      co.addCancellableTask("a", "task0-copy0")(task0)
+      co.addCancellableTask("a", "task0-copy1")(task0)
+      val cancellable0 = co.addCancellableTask("a", "task0-copy2")(task0)
+
+      object Task1 { // We can also add a method as a task
+        val counter = new AtomicInteger()
+
+        private def taskMethod(): Future[Done] = Future {
+          counter.incrementAndGet()
+          Done
+        }
+
+        def register(phase: String, name: String): Cancellable = co.addCancellableTask(phase, name)(taskMethod _)
+      }
+
+      Task1.register("b", "task1-copy0")
+      Task1.register("b", "task1-copy1")
+      val cancellable1 = Task1.register("b", "task1-copy2")
+
+      // Adding the same task twice under the same phase/name will still run it twice
+      val task2counter = new AtomicInteger()
+      co.addCancellableTask("a", "task2")(() =>
+        Future {
+          task2counter.incrementAndGet()
+          Done
+        })
+      co.addCancellableTask("a", "task2")(() =>
+        Future {
+          task2counter.incrementAndGet()
+          Done
+        })
+      val cancellable2 = co.addCancellableTask("a", "task2")(() =>
+        Future {
+          task2counter.incrementAndGet()
+          Done
+        })
+
+      object TaskAB { // tests cancellation by tasks in a previous/later phase
+        val promiseA = Promise[Unit]
+        val promiseB = Promise[Unit]
+        val taskA: Cancellable = co.addCancellableTask("a", "task-a") { () =>
+          Future {
+            taskB.cancel()
+            promiseA.trySuccess(log.info("Completing Promise A"))
+            Done
+          }
+        }
+        val taskB: Cancellable = co.addCancellableTask("b", "task-b") { () =>
+          Future {
+            taskA.cancel()
+            promiseB.trySuccess(log.info("Completing Promise B"))
+            Done
+          }
+        }
+      }
+
+      TaskAB.promiseA.isCompleted shouldBe false
+      TaskAB.promiseB.isCompleted shouldBe false
+      TaskAB.taskA.isCancelled shouldBe false
+      TaskAB.taskB.isCancelled shouldBe false
+
+      cancellable0.cancel()
+      cancellable1.cancel()
+      cancellable2.cancel()
+
+      whenReady(co.run(UnknownReason)) { _ =>
+        task0Counter.get shouldEqual 2
+        Task1.counter.get shouldEqual 2
+        task2counter.get shouldEqual 2
+        TaskAB.promiseA.isCompleted shouldBe true
+        TaskAB.promiseB.isCompleted shouldBe false
+      }
```

Comment:
I'd like to see some more tests around concurrent registration and cancellation given that is the meat of this PR. Something that adds and cancels tasks on a few thread for a not to large number of iterations and then checks that the right number of tasks is still in there.

Reply:
Done

Reply:
Good idea